urllib3/connectionpool.py

```diff
@@ -381,8 +382,22 @@ def _make_request(self, conn, method, url, timeout=_Default,
         log.debug("\"%s %s %s\" %s %s" % (method, url, http_version,
                                           httplib_response.status,
                                           httplib_response.length))
+
+        header_parsing_errors = extract_parsing_errors(httplib_response.msg)
```

Comment:
As far as the header parsing layer is concerned (urllib3.util.request), it has failed to perform its job given the input. It raises an exception.

Then the layer above is like "oh well, I don't mind if some headers are missing, I'll handle it by emitting a warning and continuing." That's how we're using exceptions everywhere, see all the socket error related exceptions that we attempt to recover from rather than outright fail.

As long as the respective layer of abstraction considers what happened a failure, exceptions are appropriate IMO.